[BEAM-65] SplittableDoFn prototype.

jkff opened the pull request:
Work in progress. Currently only runs in direct runner,
and not ready for any use by real users.

No tests for most of the code yet (will add later in this PR) but the rest is ready to take a look.

jkff commented:
R: @bjchambers  
R: @tgroh 
CC: @dhalperi , @kennknowles , @dpmills

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/GroupByKeyIntoKeyedWorkItems.java:
```diff
@@ -0,0 +1,40 @@
+ * Interface for creating a runner-specific {@link GroupByKey}-like {@link PTransform} that produces
+ * {@link KeyedWorkItem} in order to give the downstream transform access to timers.
+ */
+public interface GroupByKeyIntoKeyedWorkItems<KeyT, InputT> {
```
I think the general approach for transforms like this would be to introduce a primitive PTransform (GroupByKeyIntoKeyedWorkItems) and then have each runner override that in their translator by providing an appropriate implementation. This is particularly important as we move towards creating a graph and then applying the appropriate runner-specific transforms -- we may want to be able to create a graph that has a general "GroupByKeyIntoKeyedWorkItems" before having a specific runner selected.

jkff replied:
Yes, this is exactly where this class is going - but it needs the input coder to produce the runner-specific PTransform (which, of course, can be primitive). I'm not sure how to express this by a single runner-agnostic primitive PTransfrom class - i.e. not sure how to pass the input coder?

The input coder is needed, e.g. in case of direct runner, to put it into the KeyedWorkItem coder: the input is PCollection&lt;KV&lt;String, InputT&gt;&gt; and to create a KeyedWorkItemCoder we need the String and InputT coders separately. I didn't want to downcast to a KvCoder :-/

tgroh replied:
For something like GroupByKey we do make the restriction that the input coder must be a KeyCoder; It's not unreasonable to do the same here.

https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupByKey.java#L239

jkff replied:
Done, good point, this allowed to easily use a primitive transform.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,399 @@
+ * restrictions, assign unique key and group by it (in order to get access to per-key state and
+ * timers), then process.
+ */
+public class SplittableParDo<
```
Ideally, the name and Javadoc would clarify that this is *not* a user-visible transform. Instead, it is a suggested implementation for a `ParDo.of(SplittableDoFn)`.

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,399 @@
+   * by creating a tracker for the restriction and checkpointing/resuming processing later if
+   * necessary.
+   *
+   * <p>TODO: This uses deprecated OldDoFn since DoFn does not provide access to state/timer
```
Specifically, this should be revisited once the State API is available to new DoFns.

jkff replied:
Done.

tgroh replied:
To enable dynamic work rebalancing and give runners more control into the execution of a split, this seems as though it might be better to rewrite in a manner similar to `ReduceFnRunner` that exercises the underlying APIs directly, but exposes additional handles to enable the runner to more richly interact with the abstractions (especially `RestrictionTracker` similar to the current `Reader/Source` APIs)

bjchambers replied:
Not entirely convinced -- we should figure out how Splittable DoFn should be exposed to runners as part of the Fn API, and implement something mimicing that. If all runners will use this expansion and provide state, this is fine. If runners may interact with the Fn API to execute the SDF differently, then we may need to refactor towards that API (by extracting the actual SDF runner pieces from this utility PTransform).

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,399 @@
+        // smoothly, and ensures that not too much work will have to be reprocessed in the event of
+        // a crash.
+        // TODO: Also commit at least once every N seconds (runner-specific parameter).
+        private static final int MAX_OUTPUTS_PER_BUNDLE = 10000;
```
Seems like 10k should be a runner parameter. For a streamier runner, that may be too much.

jkff replied:
Yeah, I'm going to change the way a runner hooks into checkpointing when I start work on Dataflow runner support. It's a bigger change than I'd like to make in this PR.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,399 @@
+      return new DoFn.ExtraContextFactory<InputT, OutputT>() {
+        @Override
+        public BoundedWindow window() {
+          return null;
```
Should these throw (to indicate they're not supported)?

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,399 @@
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      for (RestrictionT part :
+          invoker.invokeSplitRestriction(c.element().getKey(), c.element().getValue(), 0)) {
```
Assign 0 to a constant to indicate what it represents?

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -306,6 +311,8 @@ public Duration getAllowedTimestampSkew() {
      */
     OutputReceiver<OutputT> outputReceiver();
+
+    <RestrictionT> RestrictionTracker<RestrictionT> restrictionTracker();
```
Javadoc on method

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -372,14 +385,57 @@ public BoundedWindow window() {
+   *     the {@link RestrictionTracker} argument of {@link ProcessElement}, which in turn must be a
+   *     subtype of {@code RestrictionTracker<R>} where {@code R} is the restriction type returned
+   *     by {@link GetInitialRestriction}.
+   * <li>It <i>must</i> define a {@link GetRestrictionCoder} method.
```
Couldn't the RestrictionCoder can't be part of the RestrictionTracker? If RestrictionTracker had a `getRestrictionCoder` method that received the input/output coders, it could encapsulate the logic to encode a restriction of type `R`, correct? That would reduce the burden on each `DoFn` using that `RestrictionTracker`.

jkff replied:
RestrictionTracker is associated with an active processElement call, but the coder is needed in contexts where no such call is in progress. I updated this documentation to explain that coder inference still applies. E.g. the example in SplittableDoFnTest doesn't define a GetRestrictionCoder method.

(Idea for future: inference for other methods, e.g. have a default newTracker and splitRestriction methods depending on restriction type)

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -372,14 +385,57 @@ public BoundedWindow window() {
+   *     current element.
    * </ul>
+   *
+   * <h2>Splittable DoFn's</h2>
```
Add a big note here about the experimentalness of the feature?

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -415,6 +471,140 @@ public BoundedWindow window() {
+   * Annotation for the method that maps an element to an initial restriction for a <a
+   * href="https://s.apache.org/splittable-do-fn">splittable</a> {@link DoFn}.
+   *
+   * <p>Signature: {@code RestrictionT getInitialRestriction(InputT element);}
```
Is the `InputT` optional? What if the `RestrictionT` is independent of it?

Also, I like this trend of including the `Signature` section in each annotation.

jkff replied:
I'd prefer not to clutter this PR with more convenience code such as optional parameters - I'd rather leave that for later when the rest of it works.

bjchambers replied:
Then add a TODO to make sure we address these later?

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -415,6 +471,140 @@ public BoundedWindow window() {
+   * <p>Signature: {@code List<RestrictionT> splitRestriction(RestrictionT restriction, int
+   * numParts);}
+   *
+   * <p>{@code numParts} is a hint as to the number of parts into which the restriction should be
```
Is it also allowed to not take the numParts parameter?

jkff replied:
Same reply as above.

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -415,6 +471,140 @@ public BoundedWindow window() {
+  public static class ProcessContinuation {
+    /** Indicates that there is no more work to be done for the current element. */
+    public static ProcessContinuation stop() {
+      return null;
```
I think I'd prefer there be a specific instance which corresponds to stop. Consider:

```
private static final ProcessContinuation STOP = new ProcessContinuation(null, null);
public static ProcessContinuation stop() {
  return STOP;
}
```

Specifically, I worry that null could be unintentional, while ProcessContinuation.stop() is nice and explicit.

jkff replied:
Done.

tgroh commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,403 @@
+        .setCoder(splitCoder)
+        .apply("Split restriction", ParDo.of(new SplitRestrictionFn<InputT, RestrictionT>(fn)))
+        .apply(
+            "Assign unique key", ParDo.of(new AssignRandomUniqueKeyFn<KV<InputT, RestrictionT>>()))
```
Do restrictions have any sort of identifier? Is it reasonable to require them to do so? If so, key by that.

jkff replied:
I don't think we should require restrictions to have an ID. They're arbitrary simple mathematical objects, we can generate an ID ourselves without requiring the user to.

tgroh commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,403 @@
+                    fn,
+                    input.getCoder(),
+                    input.getWindowingStrategy().getWindowFn().windowCoder())))
+        .setIsBoundedInternal(isFnBounded);
```
This should be `input.isBounded().and(isFnBounded)`; any unboundedness propagates

jkff replied:
Done.

tgroh commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,403 @@
+      TimerInternals timerInternals = internals.timerInternals();
+
+      ValueState<WindowedValue<InputT>> elementState =
+          stateInternals.state(StateNamespaces.global(), elementTag);
```
(Assuming I understand the namespaces correctly), should the PTransform also ensure that its input window fn is `GlobalWindows`?

bjchambers replied:
I'm not sure it matters -- the namespace just means that we're getting key-only state tags rather than a state tag associated with the window.

That said, it does raise a question -- if there are multiple input elements on a single key, this would break down. But, I think that we've rekeyed so that every element is on a unique key. We should document that as a precondition, and possibly mention it on the StateTags.

We may want to (in the future) discuss adding a state namespace which is just "unique for this element" or something.

jkff replied:
I could use c.window() after https://github.com/apache/incubator-beam/pull/863 is merged. Currently I get an exception from c.window() if this is a processElement() call for a timer firing.

However, with that fix, during a timer firing I'll get the global window from c.window(), and then I'll be looking up the state in the wrong window if the original element was not in the global window.

I was wondering if maybe I should get the window from the fired timer's namespace? But this seems awkward to do as well because the KeyedWorkItem has multiple timers (timersIterable()) - hmm, should I process them each as if they were separate KeyedWorkItem's?

bjchambers replied:
You're only setting one timer per element, right? So you should never see multiple timers on a single key, correct?

jkff replied:
Right, that helped!

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -87,17 +102,159 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
+                ? PCollection.IsBounded.UNBOUNDED
+                : PCollection.IsBounded.BOUNDED;
+      }
+      checkNotNull(
```
It would be nice to condense these exceptions to throw once with something like

"... defines @ProcessElement method ... which is splittable, but does not define required methods %s, missingMethods"

so you get the failure once, and then can fix everything without having to run your pipeline.

jkff replied:
Done, I changed the error reporting quite a bit.

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -87,17 +102,159 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
+          format(processElementMethod));
+      isBounded = PCollection.IsBounded.BOUNDED;
+
+      checkArgument(
```
Likewise with condensed messaging

jkff replied:
Ditto.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,403 @@
+    }
+
+    @Override
+    public void processElement(final ProcessContext c) {
```
This `processElement` method is long. It would be good to split it up into some helper functions or even a helper class to make the different steps clearer.

bjchambers replied:
Seems reasonable for now. As this becomes more complex (for other runners) we may want to introduce a `SplittableDoFnRunner` or something like that. The benefit is that common parameters can be assigned in the constructor, it can build up state if needed, etc. which should make it easier to split up. (eg., this is what we did for ReduceFnRunner, etc.). It also allows unit testing the life-cycle logic directly.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,403 @@
+
+        private void noteOutput() {
+          if (++numOutputs > MAX_OUTPUTS_PER_BUNDLE) {
+            // Request a checkpoint. The fn *may* produce more output, but hopefully not too much.
```
I wonder if we should at some point add a log or a counter to track how much than it produces. It could be a good measure of how strictly splittable a given fn is?

jkff replied:
I think tracking the amount of output produced by a DoFn is not specific to splittable DoFn's.
If you mean tracking the amount of *potential parallelism* (number of claimed blocks), then that's number of .tryClaim() calls on the RestrictionTracker and I'd rather address this much later because it's in the "limited parallelism" signal territory and we haven't yet figured out how to cleanly do it in SDF.

bjchambers replied:
I think I meant tracking how much more than the requested split was produced. It can definitely be future work. But, for an arbitrary DoFn, it would be good to know that if we asked for a checkpoint at 10000 elements, if we got 10000, 10001, or 500000 (one of these is less finely splittable than the others).

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,403 @@
+      return new DoFn.ExtraContextFactory<InputT, OutputT>() {
+        @Override
+        public BoundedWindow window() {
+          throw new UnsupportedOperationException("Splittable DoFn's don't support extra context");
```
And in fact this shouldn't be reachable, since the user wasn't able to ask for any of these, correct?

jkff replied:
Done, throwing different exception.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,403 @@
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      for (RestrictionT part :
```
I'd suggest putting the result of `invokeSplitRestriction` into a local variable to improve readability of the loop.

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java:
```diff
@@ -310,78 +357,100 @@ protected StackManipulation invokeTargetMethod(MethodDescription instrumentedMet
-   * Implements {@link DoFnInvoker#invokeStartBundle} or {@link DoFnInvoker#invokeFinishBundle} by
-   * delegating respectively to the {@link StartBundle} and {@link FinishBundle} methods.
+   * Delegates to the given method, wrapping the call into a UserCodeException and optionally
+   * reparing parameters.
```
"reparing" (I assume repairing) and also not sure exactly what that means. How are parameters damaged?

jkff replied:
I have no idea what I had in mind. Fixed.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java:
```diff
@@ -310,78 +357,100 @@ protected StackManipulation invokeTargetMethod(MethodDescription instrumentedMet
    * Wraps a given stack manipulation in a try catch block. Any exceptions thrown within the try are
    * wrapped with a {@link UserCodeException}.
+   *
+   * <p>Assumes that the stack manipulation contains a method return at the end.
```
I'm guessing the assumption led to the simplification below? I'm a bit worried that it may not apply to all cases. Was there a reason we needed to remove the catchBlockEnd to make this PR work?

jkff replied:
Honestly the reason was, I spent about a day trying to make it work without this assumption, and kept getting obscure bytecode verification errors that I was unable to interpret (and I dug pretty deep into the JVM spec). So I figured I'd just do it like this for now, and fix when this assumption is violated.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignature.java:
```diff
@@ -32,65 +37,131 @@
+  @Nullable
+  public abstract NewTrackerMethod newTracker();
+
   static DoFnSignature create(
```
In this or a follow-up CL, I would strongly suggest switching DoFnSignature to using a builder rather than a single create method. This is getting pretty long, and especially since it has multiple nullable fields, etc., it would be really easy for us to mix them up with odd consequences.

jkff replied:
Done, used builder.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -65,13 +68,25 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
+      }
+      if (supertype.getRawType().isAnnotationPresent(DoFn.Unbounded.class)) {
+        checkArgument(
+            isBounded == null, "Both @Bounded and @Unbounded specified on %s", formatType(fnToken));
```
I think at some point we'll want this to use machinery like `AvroDeterminismChecker` (https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java#L362) which encapsulates the validation code and allows recording multiple problems before failing rather than only reporting the first problem.

That should probably be done in a separate PR.

jkff replied:
Agreed that this is better done in separate PR.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -605,4 +553,38 @@ private static String format(Method m) {
     return (t == null) ? "null" : ReflectHelpers.TYPE_SIMPLE_DESCRIPTION.apply(t.getType());
   }
+
+  static class ErrorReporter {
```
1. I'd propose adjusting so that you can always get the root error reporter (since this will support marking failures etc. in the future).
2. I'd also consider just maintaining the chain as you go down (in nest) rather than having the labelChain approach.
3. Rather than a generic "nest" I'd make one or two helpers that are applicable to the kinds of nesting we need. This allows us to have one place that deals with formatting of nesting (this class) rather than having it spread out via String.format.

Something like:

```
static class ErrorReporter {
 private final ErrorReporter root;
  private final String prefix;
  private ErrorReporter(ErrorReporter root, String prefix) {
    this.root = root;
    this.prefix = prefix;
  }
  public ErrorReporter(String prefix) {
    this.root = this;
    this.prefix = prefix;
  }
  public ErrorReporter forFunction(String annotation, Method m) {
    return new ErrorReporter(this.root, String.format(...));
  }
}
```

jkff replied:
Simplified; now neither a parent nor root reporter is necessary. Might add a root reporter when it becomes necessary (gathering multiple errors).

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java:
```diff
@@ -431,18 +498,6 @@ public Size apply(MethodVisitor mv, Implementation.Context implementationContext
-        mv.visitLabel(catchBlockEnd);
-        // The frame contents after the try/catch block is the same
-        // as it was before.
-        mv.visitFrame(
```
I think that removing this will create problems for the exception stack traces (could be wrong). I'm guessing this is what was giving you problems -- something about the frame (which is used for stack unwinding) disagreed with the code that was added in the block? Maybe there were some local variables/stack variables added within the block that needed to be unwound?

jkff replied:
Merged your PR that fixes this.

tgroh commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -208,71 +212,69 @@ public void setup() throws Exception {
-      final boolean wasFirstCall = (restriction == null);
-      if (wasFirstCall) {
-        // This is the first ProcessElement call for this element/restriction.
+      ElementRestriction<WindowedValue<InputT>, RestrictionT> state;
```
this seems like it should be called `restriction`, given the accesses you're making to `StateInternals` (`elementState`, `restrictionState`...)

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java:
```diff
@@ -501,8 +493,8 @@ protected StackManipulation beforeDelegation(MethodDescription instrumentedMetho
-    public UserCodeMethodInvocation(
-        Integer returnVarIndex,
+    UserCodeMethodInvocation(
+        @Nullable Integer returnVarIndex,
```
This may be be better written using `Optional<Integer>` since it makes the contract more obvious.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/ElementRestriction.java:
```diff
@@ -0,0 +1,32 @@
+import com.google.auto.value.AutoValue;
+
+@AutoValue
+public abstract class ElementRestriction<ElementT, RestrictionT> {
```
Javadoc the class and methods

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/ElementRestrictionCoder.java:
```diff
@@ -0,0 +1,59 @@
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CustomCoder;
+
+public class ElementRestrictionCoder<ElementT, RestrictionT>
```
Javadoc public classes and methods

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/GBKIntoKeyedWorkItems.java:
```diff
@@ -0,0 +1,37 @@
+import org.apache.beam.sdk.values.PCollection;
+
+/**
+ * Interface for creating a runner-specific {@link GroupByKey}-like {@link PTransform} that produces
```
`{@link GroupByKey GroupByKey-like}` will javadoc better

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/GBKIntoKeyedWorkItems.java:
```diff
@@ -0,0 +1,37 @@
+
+/**
+ * Interface for creating a runner-specific {@link GroupByKey}-like {@link PTransform} that produces
+ * {@link KeyedWorkItem} in order to give the downstream transform access to timers.
```
`{@link KeyedWorkItem KeyedWorkItems}`

bjchambers replied:
is "in order" meant to imply ordered, or is it just "so that downstream transforms can access state and timers"?

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,406 @@
+ * <p>This transform is intended as a helper for internal use by runners when implementing {@code
+ * ParDo.of(splittable DoFn)}, but not for direct use by pipeline writers.
+ */
+public class SplittableParDo<
```
`@Experimental` on this and other public classes?

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,406 @@
+   * @param fn The splittable {@link DoFn} inside the original {@link ParDo} transform.
+   */
+  public SplittableParDo(String name, DoFn<InputT, OutputT> fn) {
+    super(name);
```
I don't think we need to add support for naming this transform. We're mostly moved away from that and on to using the named apply (`.apply(name, transform)` rather than `.apply(transform.named(name))`).

jkff replied:
Done.

bjchambers commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoOverrideFactory.java:
```diff
@@ -0,0 +1,55 @@
+    if (!signature.processElement().isSplittable()) {
+      return transform;
+    }
+    return new SplittableParDo(that.getName(), fn);
```
Should also make sure that other runners clearly fail on if they receive a splittable `DoFn`, since this will only be supported by the `DirectRunner` for now. Eg., maybe the `PTransform` should fail at apply/validate time if it receives a splittable `DoFn`?

jkff replied:
Done. ParDo.apply() now fails if the DoFn is splittable.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,406 @@
+   * collection is effectively the same elements as input, but with access to per-element (per-key)
+   * state and timers.
+   */
+  private static class AssignRandomUniqueKeyFn<T> extends DoFn<T, KV<String, T>> {
```
This could use `WithKeys` I believe, but it's fine to use a specific transform and better documentation (like this).

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,406 @@
+     * DoFn.ProcessContinuation#stop}.
+     *
+     * <p>A hold is needed to avoid letting the output watermark immediately progress together with
+     * the input watermark when the first {@link DoFn.ProcessElement} call completes.
```
after the first split of the element completes?

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,406 @@
+    @Override
+    public void setup() throws Exception {
+      invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
+      restrictionTag =
```
I don't see why this tag isn't also created in the constructor?

jkff replied:
Done. (note that field "invoker" is transient so I can't initialize invoker in the constructor)

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,406 @@
+        }
+
+        public <T> void sideOutput(TupleTag<T> tag, T output) {
+          // TODO: I'm not sure this is correct, but there's no "internals.sideOutputWindowedValue".
```
Maybe just `throw UnsupportedOperationException("Side outputs not yet supported by SplittableDoFn")`? Which also raises the question, what if the `DoFn` produces more side-outputs than main-outputs? Shouldn't this increase the output counter too?

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,406 @@
+
+        public <T> void sideOutputWithTimestamp(TupleTag<T> tag, T output, Instant timestamp) {
+          // TODO: I'm not sure this is correct, but there's no "internals.sideOutputWindowedValue".
+          baseContext.sideOutputWithTimestamp(tag, output, timestamp);
```
Ditto re:throwing until we properly support it?

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,406 @@
+      return new DoFn.ExtraContextFactory<InputT, OutputT>() {
+        @Override
+        public BoundedWindow window() {
+          throw new IllegalStateException("Unexpected extra context access on a splittable DoFn");
```
Maybe add a comment that says the `DoFnReflector` or whatever should ensure this never happens?

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,406 @@
+
+        @Override
+        public DoFn.InputProvider<InputT> inputProvider() {
+          throw new IllegalStateException("Unexpected extra context access on a splittable DoFn");
```
.

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,406 @@
+
+        @Override
+        public DoFn.OutputReceiver<OutputT> outputReceiver() {
+          throw new IllegalStateException("Unexpected extra context access on a splittable DoFn");
```
.

jkff replied:
Done.

bjchambers commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectGBKIntoKeyedWorkItemsOverrideFactory.java:
```diff
@@ -0,0 +1,62 @@
+import org.apache.beam.sdk.values.PInput;
+import org.apache.beam.sdk.values.POutput;
+
+class DirectGBKIntoKeyedWorkItemsOverrideFactory implements PTransformOverrideFactory {
```
Javadoc (it's still useful on top-level classes that aren't public ;))

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -372,14 +390,57 @@ public BoundedWindow window() {
-   *   <li>Its first argument must be a {@link DoFn.ProcessContext}.
-   *   <li>Its remaining argument, if any, must be {@link BoundedWindow}.
+   * <li>Its first argument must be a {@link DoFn.ProcessContext}.
+   * <li>Its remaining arguments, if any, must be either {@link BoundedWindow} or a subtype of
```
Thinking about maintaining this moving forward, I wonder if it would be easier to have something like:

"If it takes a {@link RestrictionTracker} as an argument, then it is a <a href...>splittable</a> DoFn and is subject to the requirements specified {@link below section}. Otherwise, the following restrictions apply:"

That way we don't have to merge all SDF parameter notes into this, and maintain them seprately (eg., this sentence seems like like will become difficult as we add more arguments to both, etc.)

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -415,6 +476,167 @@ public BoundedWindow window() {
+   *
+   * <p>{@code numParts} is a hint as to the number of parts into which the restriction should be
+   * split. The method is allowed to respect the hint or to ignore it and produce any number of
+   * parts. 0 means "no hint, choose a reasonable value".
```
`UNSPECIFIED_NUM_PARTS means "no hint, choose a reasonable value"

jkff replied:
I removed the parameter altogether because the only supporting runner set it only to the meaningless value. I'll reintroduce it together with support for passing something non-trivial from a runner.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -415,6 +476,167 @@ public BoundedWindow window() {
+  @Retention(RetentionPolicy.RUNTIME)
+  @Target(ElementType.TYPE)
+  @Experimental(Kind.SPLITTABLE_DO_FN)
+  public @interface Bounded {}
```
What if we used the presence/absence of one of these to drive SDF? Eg., "BoundedSDF" or "UnboundedSplittableDoFn"? It would be easier to document a lot of things:

`If the DoFn is annotated as a {@link BoundedSplittableDoFn bounded} or @{link UnboundedSplittableDoFn} it has these requirements ... Otherwise, it has these requirements ...` etc.?

bjchambers replied:
Also it should be an error to have these on a non-splittable DoFn.

jkff replied:
I don't want to rename to "BoundedSDF" because stateful DoFn's, in the future, can also be bounded/unbounded. It seems like a generic enough annotation, even though right now it is specific to SDF.
This change already makes it an error to specify these on a non-splittable DoFn. When we have stateful fns, we'll need to fix the doc and code to say it's also allowed on a stateful fn.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -415,6 +476,167 @@ public BoundedWindow window() {
+    }
+
+    private final Duration resumeDelay;
+    @Nullable private final Instant futureOutputWatermark;
```
Rather than nullable, may be clearer if this was `Optional.present(...)` to indicate a user requested future output watermark, and `Optional.absent()` to indicate no preference. Then it is explicit rather than "if its null, then its the default, which should be ..."

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvoker.java:
```diff
@@ -43,7 +46,23 @@
+  /** Invoke the {@link DoFn.GetRestrictionCoder} method on the bound {@link DoFn}. */
+  <RestrictionT> Coder<RestrictionT> invokeGetRestrictionCoder();
+
+  /** Invoke the {@link DoFn.SplitRestriction} method on the bound {@link DoFn}. */
```
Would probably be good to document these more. Eg., what should a runner pass as "numParts", etc.

Eg., write these as documentation for runners and such that will be invoking `DoFn` methods, and the Javadoc in `DoFn` is more for people who are writing DoFns.

jkff replied:
Removed the parameter. Other methods seem to not require more explanation.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -213,16 +353,34 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
+    if (extraParameters.contains(DoFnSignature.Parameter.RESTRICTION_TRACKER)) {
+      errors.checkArgument(
+          extraParameters.size() == 1,
+          "Splittable and must not have any extra context arguments apart from %s, but has: %s",
```
"Splittable DoFns must not have ..."

bjchambers commented:
Haven't looked at tests, but seems generally reasonable as a first checkin

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/GBKIntoKeyedWorkItems.java:
```diff
@@ -0,0 +1,40 @@
+ * and timers.
+ */
+@Experimental(Experimental.Kind.SPLITTABLE_DO_FN)
+public class GBKIntoKeyedWorkItems<KeyT, InputT>
```
Mostly as an FYI for @kennknowles -- the State API should affect this part of the implementation.

kennknowles replied:
This PR is stretching the limits of GitHub's UI. Can you point me to where this is used?

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,412 @@
+/**
+ * A utility transform that executes a <a
+ * href="https://s.apache.org/splittable-do-fn">splittable</a> {@link DoFn} by expanding it into a
+ * network of simpler transforms: pair with restriction, split restrictions, assign unique key and
```
```
<ol>
<li>Pair each element with a restriction tracker
<li>Split all of the restrictions
<li>Assign a unique key to each split of each element
<li>Group by key (so that work is partitioned by key and we can access state/timers)
<li>Execute the splittable DoFn using state and timers
</ol>
```

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,412 @@
+   *
+   * @param fn The splittable {@link DoFn} inside the original {@link ParDo} transform.
+   */
+  public SplittableParDo(DoFn<InputT, OutputT> fn) {
```
```
checkNotNull(fn, "fn must not be null");
checkArgument(fn.isSplittable(), "fn must be a splittable DoFn");
```

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,412 @@
+            ParDo.of(new PairWithRestrictionFn<InputT, OutputT, RestrictionT>(fn)))
+        .setCoder(splitCoder)
+        .apply("Split restriction", ParDo.of(new SplitRestrictionFn<InputT, RestrictionT>(fn)))
+        .setCoder(splitCoder)
```
nit: This call to setCoder may not be necessary since the input type and output type are the same it should be propagated. That said, likely no harm in setting it.

jkff replied:
It's necessary for some reason - I didn't dig deeper.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,412 @@
+  /**
+   * Assigns a random unique key to each element of the input collection, so that the output
+   * collection is effectively the same elements as input, but with access to per-element (per-key)
+   * state and timers.
```
"effectively the same elements as input, but the per-key state and timers are now effectively per-element"

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,412 @@
+      implements OldDoFn.RequiresWindowAccess {
+    /**
+     * The state cell containing a watermark hold for the output of this {@link DoFn}. The hold is
+     * acquired during the first {@link DoFn.ProcessElement} call for this element and restriction,
```
"for each element and restriction, and is released when ..."

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,412 @@
+    }
+
+    @Override
+    public void processElement(final ProcessContext c) {
```
I think this could be cleaned up with something like:

1. A class that holds all the details necessary for running one of these. This could just be a simple AutoValue holder class.
2. Methods for creating this data from a seed, and a method for populating this from existing information.

Something like:

```
ElementData data = (timer == null ? createSeedData(c) : loadData(c));
```

Where this `ElementData` has the various states necessary, and any other information. This should also make it easier to split this method into pieces,  since different parts can just interact with the `ElementData`.

jkff replied:
I spent a few hours trying to do this and couldn't come up with anything remotely clean. This will change a lot in subsequent PRs where I'll add runner hooks into the heart of this method. I'd rather keep it like this for now; I think a linearly structured method of this size is not so bad, and it was definitely more readable than anything else I was able to come up with as a decomposition.

bjchambers commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoOverrideFactory.java:
```diff
@@ -0,0 +1,55 @@
+    if (!signature.processElement().isSplittable()) {
+      return transform;
+    }
+    return new SplittableParDo(fn);
```
In theory, we could pass the DoFnSignature into the SplittableParDo since it is useful there.

jkff replied:
We could, but I don't like the idea of passing both the fn and a signature into SplittableParDo, since one can be produced from the other (so the 1-argument API seems simpler and less error-prone), and we don't yet have evidence of a performance problem here.

bjchambers replied:
@kennknowles may be playing with similar parts of the code while switching to new DoFn, so FYI.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -167,7 +304,10 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
+    errors.checkArgument(
+        void.class.equals(m.getReturnType())
+            || DoFn.ProcessContinuation.class.equals(m.getReturnType()),
+        "Must return void or ProcessContinuation");
```
In general, I'd suggest `"Must return void or %s", ProcessContinuation.class.getSimpleName()` since it is more stable under IDE renamings.

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -213,16 +354,34 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
+      } else if (RestrictionTracker.class.isAssignableFrom(rawType)) {
+        errors.checkArgument(
+            !extraParameters.contains(DoFnSignature.Parameter.RESTRICTION_TRACKER),
+            "Multiple RestrictionTracker parameters");
```
Ditto here and else-where that class-names are used.

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -120,6 +140,123 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
+      }
+      processElementErrors.checkArgument(
+          processElement.trackerT().equals(newTracker.trackerT()),
+          "Has tracker type %s, but @NewTracker method %s uses tracker type %s",
```
Same comment regarding `@%s` + `NewTracker.class.getSimpleName()` here and throughout (for forbidden methods, required methods, etc.)

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokersTest.java:
```diff
@@ -292,6 +417,22 @@ public void processElement(@SuppressWarnings("unused") ProcessContext c) {
   }
 
   @Test
+  public void testProcessElementExceptionWithReturn() throws Exception {
```
Do we have test that use an input of type `long[]` (just for sanity checking things)

jkff replied:
I don't think we use type of input and output of the DoFn as type of local variable, anything on the stack, or return type in generated code. Why do we need such a test?

tgroh commented on runners/direct-java/src/test/java/org/apache/beam/runners/direct/SplittableDoFnTest.java:
```diff
@@ -0,0 +1,145 @@
+    }
+  }
+
+  private static class OffsetRangeTracker implements RestrictionTracker<OffsetRange> {
```
TODO: Thread safety?

jkff replied:
In this PR, this is not yet accessed in a concurrent fashion; and this is a test class. I'll update this accordingly in future PRs; and of course when I add OffsetRangeTracker as a utility intended for consumption by users, it'll be thread-safe and the one in this test will be removed.

jkff commented:
OK, I rebased, added unit tests, and addressed most comments. I'll address the couple more final comments meanwhile, but as far as I'm concerned this PR should be good to go modulo cosmetics.

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -372,74 +345,252 @@ public BoundedWindow window() {
+     * A minimum duration that should elapse between the end of this {@link ProcessElement} call and
+     * the {@link ProcessElement} call continuing processing of the same element. By default, zero.
+     */
+    public abstract boolean shouldResume();
```
Done.

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -372,74 +345,252 @@ public BoundedWindow window() {
+     * <p>A runner should treat an absent value as equivalent to the timestamp of the input element.
+     */
+    @Nullable
+    public abstract Instant futureOutputWatermark();
```
Done.

On your next comment (for some reason I can't reply to it there): I will do that in subsequent PRs that add runner hooks and support for Dataflow runner. I don't yet have enough information on what's the right way to do it in this PR.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/ElementRestriction.java:
```diff
@@ -0,0 +1,42 @@
+ */
+@Experimental(Experimental.Kind.SPLITTABLE_DO_FN)
+@AutoValue
+public abstract class ElementRestriction<ElementT, RestrictionT> {
```
Could this be `ElementAndRestriction` or `ElementWithRestriction`? As described, this made me ask "how does this restrict an element" and that lead to "what is this tuple" later on.

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/ElementRestriction.java:
```diff
@@ -0,0 +1,42 @@
+  /** The restriction applied to processing the element. */
+  public abstract RestrictionT restriction();
+
+  /** Constructs the tuple. */
```
Constructs the `{@link ElementRestriction}`?

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/ElementRestrictionCoder.java:
```diff
@@ -0,0 +1,63 @@
+    ElementT key = elementCoder.decode(inStream, nestedContext);
+    RestrictionT value = restrictionCoder.decode(inStream, nestedContext);
+    return ElementRestriction.of(key, value);
+  }
```
Should this support `verifyDeterimnistic` so we can use this as a key coder? Will that ever be necessary?

jkff replied:
I don't think it's necessary. ElementAndRestriction is an internal detail of SplittableParDo and I don't see why SplittableParDo would use it as a key - if it does, we'll fix this.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/ElementRestrictionCoder.java:
```diff
@@ -0,0 +1,63 @@
+      throws IOException {
+    Context nestedContext = context.nested();
+    elementCoder.encode(value.element(), outStream, nestedContext);
+    restrictionCoder.encode(value.restriction(), outStream, nestedContext);
```
I think technically we could encode one or the other (probably the element) using the original context, which may save some space. For instance:

```
restrictionCoder.encode(vaule.restriction(), outStream, context.nested());
elementCoder.encode(value.element(), outStream, context);
```

jkff replied:
Tried this and the other symmetric variant, it didn't work. The current code matches KvCoder.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/ElementRestrictionCoder.java:
```diff
@@ -0,0 +1,63 @@
+
+/** A {@link Coder} for {@link ElementRestriction}. */
+@Experimental(Experimental.Kind.SPLITTABLE_DO_FN)
+public class ElementRestrictionCoder<ElementT, RestrictionT>
```
This new class should have the suite of Coder tests that are part of CoderProperties.

jkff replied:
Done. Fixed KvCoderTest while I was at it.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,435 @@
+                "Group by key",
+                new GBKIntoKeyedWorkItems<String, ElementRestriction<InputT, RestrictionT>>());
+    checkArgument(
+        keyedWorkItems.getWindowingStrategy().getWindowFn() instanceof GlobalWindows,
```
I think we could move this earlier -- we could check the windowing strategy of the input and/or check this in validate (rather than apply).

Everything within here is preserving the original windowing.

bjchambers replied:
And below we even ignore the keyedWorkItems windowing and go back to the input.getWindowingStrategy().

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,435 @@
+   * collection is effectively the same elements as input, but the per-key state and timers are now
+   * effectively per-element.
+   */
+  private static class RandomUniqueKeyFn<T> implements SerializableFunction<T, String> {
```
Should this use the new transform to assign keys without using UUID? Should that be a TODO?

jkff replied:
No, that transform (I suppose you're referring to part of Redistribute transform) is best-effort-unique-ish; the one here has to actually be globally unique.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,435 @@
+
+    /** Creates an {@link DoFn.ExtraContextFactory} that provides just the given tracker. */
+    private DoFn.ExtraContextFactory<InputT, OutputT> wrapTracker(final TrackerT tracker) {
+      return new DoFn.ExtraContextFactory<InputT, OutputT>() {
```
I generally prefer "private static class" rather than creating these as anonymous inner classes, partly because it makes it clearer what is getting captured from the outside environment (here and above).

jkff replied:
To previous comment (github doesn't let me reply there...): again, I'd rather make big restructurings of how this interacts with the runner in a separate PR - the current PR just makes the semantics work properly.

To current comment: done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,435 @@
+  private static class SplitRestrictionFn<InputT, RestrictionT>
+      extends DoFn<
+          ElementRestriction<InputT, RestrictionT>, ElementRestriction<InputT, RestrictionT>> {
+    private final DoFn<InputT, ?> fn;
```
maybe call this splittableFn or something to make it clear which `fn` we're referring to?

jkff replied:
Done.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -0,0 +1,435 @@
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      List<RestrictionT> parts =
```
Should we worry about the size of this list? If so, would it make sense to call wire up the invocation of the split restriction to immediately call output?

bjchambers replied:
I think we could do that by passing `invokeSplitRestriction` a `Receiver<RestrictionT>` that calls `c.output()`.

bjchambers commented on runners/core-java/src/test/java/org/apache/beam/runners/core/SplittableParDoTest.java:
```diff
@@ -0,0 +1,447 @@
+
+  private static class BoundedFakeFn extends DoFn<Integer, String> {
+    @ProcessElement
+    public void processElement(ProcessContext context, SomeRestrictionTracker tracker) {}
```
Looking at this, am I correct in assuming that Bounded vs. Unbounded is no longer an annotation but is inferred from whether we return ProcessContinuation? Otherwise, why is one of these Bounded and one Unbounded?

jkff replied:
It is inferred from return value if there's no annotation (void -> bounded, ProcessContinuation -> unbounded). However, you are allowed to be bounded even if you return a ProcessContinuation (but not reverse: you can't be unbounded if you return void). All these cases are unit-tested.

bjchambers commented on runners/core-java/src/test/java/org/apache/beam/runners/core/SplittableParDoTest.java:
```diff
@@ -0,0 +1,447 @@
+
+  @Test
+  @Category(RunnableOnService.class)
+  public void testBoundedness() {
```
Split these up into two tests -- one for the BoundedFakeFn and one for UnboundedFakeFn.

bjchambers commented on runners/direct-java/src/test/java/org/apache/beam/runners/direct/SplittableDoFnTest.java:
```diff
@@ -0,0 +1,222 @@
+ *
+ * <p>TODO: make this use @RunnableOnService.
+ */
+public class SplittableDoFnTest {
```
This should at least be `@RunWith(JUnit4.class)` or it won't run.

bjchambers commented on runners/core-java/src/test/java/org/apache/beam/runners/core/SplittableParDoTest.java:
```diff
@@ -0,0 +1,447 @@
+import org.junit.experimental.categories.Category;
+
+/** Tests for {@link SplittableParDo}. */
+public class SplittableParDoTest {
```
`@RunWith(JUnit4.class)`

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/annotations/Experimental.java:
```diff
@@ -76,6 +76,12 @@
+     * <a href="https://s.apache.org/splittable-do-fn">Splittable DoFn</a>.
+     * Do not use: API is unstable and runner support is incomplete.
+     */
+    SPLITTABLE_DO_FN,
```
I think that these experimental options are essentially invisible. If we feel strongly about "API is unstable", we could add a pipeline option that causes `ParDo.of(sdf)` to fail.

jkff replied:
Hmm, not sure what you mean here? Would you like stronger guarantees that users won't be able to apply this transform?

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java:
```diff
@@ -126,6 +135,38 @@ private DoFnInvokers() {}
+    private static final CoderRegistry CODER_REGISTRY = new CoderRegistry();
+
+    static {
+      CODER_REGISTRY.registerStandardCoders();
```
This may not be generally safe. If the user has registered a restriction coder, this won't be picked up here, for instance.

jkff replied:
Good point, fixed, plumbed CoderRegistry to here.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java:
```diff
@@ -159,7 +200,27 @@ public String subclass(TypeDescription.Generic superClass) {
+            .method(ElementMatchers.named("invokeGetRestrictionCoder"))
+            .intercept(
+                signature.processElement().isSplittable()
+                    ? ((signature.getRestrictionCoder() == null)
```
Nested ternary conditionals is the limit -- we should add a method for each of these interceptions. Something like:
`delegateSplitOrDefault(...)`
`delegateProcessElement(...)`

and then use normal if statements.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -120,6 +148,129 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
     }
 
+    DoFnSignature.GetInitialRestrictionMethod getInitialRestriction = null;
+    ErrorReporter getInitialRestrictionErrors = null;
```
Can we move these variable declarations into the if like you did with the others to be consistent?

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -120,6 +148,129 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
+
+    // Additional validation for splittable DoFn's.
+    if (processElement.isSplittable()) {
+      if (isBounded == null) {
```
Add a comment here -- this isn't validation, we're inferring the boundedness from the processElement. Should also test for both versions (annotated vs. inferred). What happens when the return value disagrees with the annotation?

jkff replied:
All cases are tested, including disagreement. Put this inference into a separate method.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -120,6 +148,129 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
+      if (newTracker == null) {
+        missingRequiredMethods.add("@" + DoFn.NewTracker.class.getSimpleName());
+      }
+      // @SplitRestriction and @GetRestrictionCoder are optional.
```
I don't know that this comment adds much, and I feel like it will quickly drift out of date. At first I didn't even get why it was here. I'd just drop it.

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -120,6 +148,129 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
+            formatType(splitRestriction.restrictionT()));
+      }
+    } else {
+      processElementErrors.checkArgument(
```
Can you split this into two methods? `verifySplittableDoFn` and `verifyNonSplittableDoFn`? That would be better than having the two big if-blocks.

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/reflect/DoFnSignaturesSplittableDoFnTest.java:
```diff
@@ -0,0 +1,520 @@
+    }
+
+    @Bounded
+    class BoundedFn extends BaseFn {}
```
Local classes tend to be iffy, but I guess this is OK. Would still be good to split this test up into InfersUnboundedness and InfersBoundedness.

bjchambers replied:
And also splitting up some of the other cases a bit more too.

jkff replied:
Split this test into several.

jkff commented:
Thanks, PTAL!

tgroh commented on runners/core-java/src/main/java/org/apache/beam/runners/core/ElementAndRestrictionCoder.java:
```diff
@@ -0,0 +1,70 @@
+      ElementAndRestriction<ElementT, RestrictionT> value, OutputStream outStream, Context context)
+      throws IOException {
+    if (value == null) {
+      throw new CoderException("cannot encode a null ElementAndRestriction");
```
This looks more like a `checkNotNull` to me, because this is a problem in the invoker's code.

jkff replied:
I mimicked KvCoder here. Not sure if I should change this - I thought coders should throw CoderException for non-encodable values. Some coders can encode a null, but this one can't.

bjchambers commented on runners/core-java/src/main/java/org/apache/beam/runners/core/ElementAndRestrictionCoder.java:
```diff
@@ -22,42 +22,49 @@
+import org.apache.beam.sdk.coders.KvCoder;
 
-/** A {@link Coder} for {@link ElementRestriction}. */
+/** A {@link Coder} for {@link ElementAndRestriction}. Parroted from {@link KvCoder}. */
```
Drop the "Parroted ..." -- this kind of thing will likely change over time.

jkff replied:
Done.
On next comment: I'm an idiot. Done too.

bjchambers commented on runners/core-java/src/test/java/org/apache/beam/runners/core/ElementAndRestrictionCoderTest.java:
```diff
@@ -0,0 +1,108 @@
+    return res;
+  }
+
+  private static final List<AnyCoderAndData> TEST_DATA =
```
If I'm reading this right, I think you can use JUnit Parameterized for this:

https://github.com/junit-team/junit4/wiki/parameterized-tests

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvoker.java:
```diff
@@ -56,11 +56,13 @@
 
   /** Invoke the {@link DoFn.GetRestrictionCoder} method on the bound {@link DoFn}. */
-  <RestrictionT> Coder<RestrictionT> invokeGetRestrictionCoder();
+  <RestrictionT> Coder<RestrictionT> invokeGetRestrictionCoder(CoderRegistry coderRegistry);
```
Add a comment about when this should be invoked. Generally, we shouldn't call methods that require a `CoderRegistry` on the worker, since it isn't available there. Maybe clarify here (and on `getRestrictionCoder` that this is called during pipeline construction)?

jkff replied:
Done on both.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java:
```diff
@@ -285,6 +280,27 @@ TrackerT invokeNewTracker(RestrictionT restriction) {
 
     final TypeDescription clazzDescription = new TypeDescription.ForLoadedType(fnClass);
 
+    final Implementation splitRestrictionDelegation;
```
Rather than the final variable + conditionals to set, why not:

```
final Implementation splitRestrictionDelegation = splitRestrictionDelegation(signature.splitRestriction())
```

And moving that to a helper method? If you want the possibilities here, I'd also see this as working:

```
splitRestrictionDelegation = Objects.firstNonNull(
    splitRestrictionFor(signature.splitRestriction()),
    MethodDelegaiton.to(DefaultSplitRestriction.class));
```

Where `splitRestrictionFor` returns null if it can't create the delegation.

jkff replied:
Done (first version - second seems less readable)

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java:
```diff
@@ -285,6 +280,27 @@ TrackerT invokeNewTracker(RestrictionT restriction) {
+          new DowncastingParametersMethodDelegation(signature.splitRestriction().targetMethod());
+    }
+    final Implementation getRestrictionCoderDelegation;
+    if (signature.processElement().isSplittable()) {
```
Same comment re: helper(s) and/or 

```
boolean isSplittable = signature.processElement().isSplittable();
final Implementation getRestrictionCoderDelegation = Objects.firstNonNull(
  isSplittable ? getRestrictionCoderFor(signatuer.getRestrictionCoder()) : null,
  isSplittable ? getDefaultRestrictionCoderFor(signatuer.getInitialRestriction()) : null,
  ExceptionMethod.throwing(UnsupportedOperationException.class));
```

jkff replied:
Also extracted into a helper but without the firstNonNull stuff.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -185,92 +168,145 @@ private static DoFnSignature parseSignature(Class<? extends DoFn<?, ?>> fnClass)
+    return signature;
+  }
+
+  private static PCollection.IsBounded inferBoundedness(
```
Some Javadoc here on the process this method uses would be helpful:

```
/**
 * Determine the boundedness of the process element method using the following criteria:
 * 1. If any of the parent types are annotated as Bounded or Unbounded, use that.
 * 2. If no annotation is present, determine boundedness from the process element return method ...
 * 3. If conflicting annotations are present, or if the method is annotated an un-splittable ...
 * etc.
 */
```

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -185,92 +168,145 @@ private static DoFnSignature parseSignature(Class<? extends DoFn<?, ?>> fnClass)
-      }
-      processElementErrors.checkArgument(
-          forbiddenMethods.isEmpty(), "Non-splittable, but defines methods: %s", forbiddenMethods);
+  private static void verifySplittableMethods(DoFnSignature signature, ErrorReporter errors) {
```
Javadoc describing the validations performed here. Eg:

```
/**
 * Validates that the following methods are present and correct ...
 * Validates that the tracker passed to processElement agrees with ...
 * etc.
 */
```

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -185,92 +168,145 @@ private static DoFnSignature parseSignature(Class<? extends DoFn<?, ?>> fnClass)
+    }
+  }
+
+  private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignature signature) {
```
Javadoc on what validations are performed.

jkff replied:
Done.

bjchambers commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvoker.java:
```diff
@@ -56,11 +56,13 @@
+  <RestrictionT> void invokeSplitRestriction(
+      InputT element,
+      RestrictionT restriction,
+      DoFn.OutputReceiver<RestrictionT> restrictionReceiver);
```
I believe `OutputReceiver` was described as "A placeholder for testing handling of output types during {@link DoFn} reflection." If we're going to use it for more, we should update that comment.

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignature.java:
```diff
@@ -35,6 +40,9 @@
   public abstract Class<? extends DoFn<?, ?>> fnClass();
 
+  /** Whether this {@link DoFn} does a bounded amount of work per element. */
+  public abstract PCollection.IsBounded isBounded();
```
I think `boundedWorkPerElement` would be clearer, despite the verbosity. I don't know what it means, otherwise, for a `DoFn` to be bounded (I do know what it means for a function to be bounded, but that is not the case here).

jkff replied:
Good point. Renamed.
Also, I considered putting it on ProcessElement method, but I think it makes more sense on the DoFn because in theory (even if not in the current API) you could imagine a DoFn that splits an element into an unbounded number of restrictions, but does bounded work in a ProcessElement call, so the total amount of work is unbounded.

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -207,20 +189,16 @@
      *
-     * <p>The element will not be changed -- it is safe to cache, etc.
-     * without copying.
+     * <p>The element will not be changed -- it is safe to cache, etc. without copying.
```
There's a lot of autoformat in this PR. Can you submit this as a prior PR to reduce the noise?

jkff replied:
Oops, reverted autoformat in DoFn. I guess I pressed the autoformat hotkey without selecting a range to format.
